src/modules/rover_mecanum/RoverMecanumControl/RoverMecanumControl.cpp

```diff
@@ -0,0 +1,220 @@
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "RoverMecanumControl.hpp"
+
+#include <mathlib/math/Limits.hpp>
+
+using namespace matrix;
+using namespace time_literals;
+
+RoverMecanumControl::RoverMecanumControl(ModuleParams *parent) : ModuleParams(parent)
+{
+	updateParams();
+	_rover_mecanum_status_pub.advertise();
+	pid_init(&_pid_yaw_rate, PID_MODE_DERIVATIV_NONE, 0.001f);
+	pid_init(&_pid_forward_throttle, PID_MODE_DERIVATIV_NONE, 0.001f);
+	pid_init(&_pid_lateral_throttle, PID_MODE_DERIVATIV_NONE, 0.001f);
+	pid_init(&_pid_yaw, PID_MODE_DERIVATIV_NONE, 0.001f);
+}
+
+void RoverMecanumControl::updateParams()
+{
+	ModuleParams::updateParams();
+	_max_yaw_rate = _param_rm_max_yaw_rate.get() * M_DEG_TO_RAD_F;
+	pid_set_parameters(&_pid_yaw_rate,
+			   _param_rm_yaw_rate_p.get(), // Proportional gain
+			   _param_rm_yaw_rate_i.get(), // Integral gain
+			   0.f, // Derivative gain
+			   1.f, // Integral limit
+			   1.f); // Output limit
+	pid_set_parameters(&_pid_forward_throttle,
+			   _param_rm_p_gain_speed.get(), // Proportional gain
+			   _param_rm_i_gain_speed.get(), // Integral gain
+			   0.f, // Derivative gain
+			   1.f, // Integral limit
+			   1.f); // Output limit
+	pid_set_parameters(&_pid_lateral_throttle,
+			   _param_rm_p_gain_speed.get(), // Proportional gain
+			   _param_rm_i_gain_speed.get(), // Integral gain
+			   0.f, // Derivative gain
+			   1.f, // Integral limit
+			   1.f); // Output limit
+	pid_set_parameters(&_pid_yaw,
+			   _param_rm_p_gain_yaw.get(),  // Proportional gain
+			   _param_rm_i_gain_yaw.get(),  // Integral gain
+			   0.f,  // Derivative gain
+			   _max_yaw_rate,  // Integral limit
+			   _max_yaw_rate);  // Output limit
+}
+
+void RoverMecanumControl::computeMotorCommands(const float vehicle_yaw, const float vehicle_yaw_rate,
+		const float vehicle_forward_speed, const float vehicle_lateral_speed)
+{
+	// Timestamps
+	hrt_abstime timestamp_prev = _timestamp;
+	_timestamp = hrt_absolute_time();
+	const float dt = math::constrain(_timestamp - timestamp_prev, 1_ms, 5000_ms) * 1e-6f;
+
+	// Update mecanum setpoint
+	_rover_mecanum_setpoint_sub.update(&_rover_mecanum_setpoint);
+
+	// Closed loop yaw control
+	if (PX4_ISFINITE(_rover_mecanum_setpoint.yaw_setpoint)) {
+		const float heading_error = matrix::wrap_pi(_rover_mecanum_setpoint.yaw_setpoint - vehicle_yaw);
+		_rover_mecanum_setpoint.yaw_rate_setpoint = pid_calculate(&_pid_yaw, heading_error, 0, 0, dt);
+
+	} else {
+		pid_reset_integral(&_pid_yaw);
+	}
+
+	// Yaw rate control
+	float speed_diff_normalized{0.f};
+
+	if (PX4_ISFINITE(_rover_mecanum_setpoint.yaw_rate_setpoint)) { 	// Closed loop yaw rate control
+		if (_param_rm_wheel_track.get() > FLT_EPSILON && _param_rm_max_thr_yaw_r.get() > FLT_EPSILON) { // Feedforward
```

Comment:
Is this check on wheel track > 0 really needed? The param has a minimum defined, that should suffice unless there are numerical instabilities that could happen.

Reply:
Agreed, removed in [df7704d](https://github.com/PX4/PX4-Autopilot/pull/23708/commits/df7704d9930a38017e7ab6bd010d7c6ae6b91555)